### PR TITLE
fix: default value not covered by parameters passed through feature file

### LINKED ISSUE #198: Function argument default values are not overridden by parameterized steps
For example:

```
@given('I have {count:w} apples')
@given('I have apples')
def given_i_have_apples(count=1):
    return count
```

The step `Given I have 5 apples` will always return 1. Is there no way to provide default values for such arguments?

### LINKED ISSUE #512: would you like reduce similar steps impl?
for this situation:

```feature
Given a phone
Given a <black> color phone
Given a <size>GB phone
Given a large dark phone
```

currently, we need write four step implements. in my ext version, we can write step impl like:

```python
@given("a phone")  # -> a white 1024GB phone
@given("a <color> color phone")  # -> a <color> 1024GB phone
@given("a <size>GB phone")  # -> a white <size> phone
@given("a large dark phone",  color="black", size=2048) # -> a black 2048GB phone
@given("a <color> color, <size>GB phone")  # -> a <color> <size> phone
def step_given_a_phone(color="white", size=1024):
    ...
```

if this feature is acceptable, i can make a PR to the latest version.

see commits:
- [support step args override](https://github.com/bulletRush/pytest-bdd/commit/9184593beabf70f4b882c9d9bc63a2423f21d575)
- [support step function default values](https://github.com/bulletRush/pytest-bdd/commit/b3c2987999a3149a7dbbde192dbbe0e21c7f8551)

### PR DESCRIPTION
**Pull request by yunxuo**
**Describe the bug**
when given multi description for step as below code

```
@given('a user with username')
@given(parsers.parse('a user with username {username}'))
@given(parsers.parse('a user with username {username} and password {password}'))
def create_user(user, username="defaultuser", password='defaultpassword'):
    user['username'] = username
    user['password'] = password
```
and pass parameters from feature file in Scenario Outline
```  
Scenario Outline: User login
    Given a user with username <username> and password <password>
    When the user logs in with username <username> and password <password>
    Then the login should be <result>

    Examples:
      | username | password    | result  |
      | testuser | password123 | success |
```
Parameters username, password parameters passed through feature file cannot take effect.
The default values in def are always used。
in above case, the result is
```
user = {
    'username': 'defaultuser',
    'password': 'defaultpassword'
}
```

see also:
https://github.com/pytest-dev/pytest-bdd/pull/610 pr fix this problems

Supersedes #610 
Resolves #512 
Resolves #198

### CODE REVIEW

**Comment by jsa34** [comment]
@yunxuo Can we add a test for this?

**Review by jsa34 — CHANGES_REQUESTED** [review]
Can we add a test for this?

**Comment by yunxuo** [comment]
> Thanks for the PR! Can we add a test for this?

yes!

**Review by jsa34 — CHANGES_REQUESTED** [review]
I cannot see any changes?  I was expecting the tests to be added

**Comment by yunxuo** [comment]
> I cannot see any changes? I was expecting the tests to be added

sorry I'm late.
I have pushed a test case in commit, but I don't know if it meets the repository specifications.

**Review by jsa34 — CHANGES_REQUESTED** [review]
I tried locally, and noticed tests fail - there seems to be issues with some step args now.

Could you address these test failures?

It appears to be when the kwarg needs to exist but isn't an arg of the step

**Comment by jsa34** [comment]
No, on the existing test.

I will invoke a run to report the tests that fail

**Comment by jsa34** [comment]
See failing tests:

https://github.com/pytest-dev/pytest-bdd/actions/runs/11854436082/job/33036622290?pr=737

**Comment by yunxuo** [comment]
> I took a look quickly to see if I could find a way to fix this and maintain the cases I mentioned, and came up with something like:
> 
> ```
> def _execute_step_function(
>     request: FixtureRequest, scenario: Scenario, step: Step, context: StepFunctionContext
> ) -> None:
>     """Execute step function."""
>     __tracebackhide__ = True
> 
>     func_sig = signature(context.step_func)
>     converters = context.converters
> 
>     def _get_parsed_arguments():
>         """Parse and convert step arguments."""
>         parsed_args = context.parser.parse_arguments(step.name)
>         if parsed_args is None:
>             raise ValueError(
>                 f"Unexpected `NoneType` returned from parse_arguments(...) in parser: {context.parser!r}"
>             )
>         kwargs = {}
>         for arg, value in parsed_args.items():
>             param = func_sig.parameters.get(arg)
>             if param:
>                 if arg in converters:
>                     value = converters[arg](value)
>                 kwargs[arg] = value
>         return kwargs
> 
>     def _get_argument_values(kwargs):
>         """Get default values or request fixture values for missing arguments."""
>         for arg in get_args(context.step_func):
>             if arg not in kwargs:
>                 param = func_sig.parameters.get(arg)
>                 if param:
>                     if param.default != param.empty:
>                         kwargs[arg] = param.default
>                     else:
>                         kwargs[arg] = request.getfixturevalue(arg)
>         return kwargs
> 
>     kw = {
>         "request": request,
>         "feature": scenario.feature,
>         "scenario": scenario,
>         "step": step,
>         "step_func": context.step_func,
>         "step_func_args": {},
>     }
> 
>     request.config.hook.pytest_bdd_before_step(**kw)
> 
>     try:
>         # Use internal methods without passing redundant arguments
>         kwargs = _get_parsed_arguments()
> 
>         if "datatable" in func_sig.parameters and step.datatable is not None:
>             kwargs["datatable"] = step.datatable.raw()
>         if "docstring" in func_sig.parameters and step.docstring is not None:
>             kwargs["docstring"] = step.docstring
> 
>         kwargs = _get_argument_values(kwargs)
> 
>         kw["step_func_args"] = kwargs
> 
>         request.config.hook.pytest_bdd_before_step_call(**kw)
> 
>         return_value = call_fixture_func(fixturefunc=context.step_func, request=request, kwargs=kwargs)
> 
>     except Exception as exception:
>         request.config.hook.pytest_bdd_step_error(exception=exception, **kw)
>         raise
> 
>     if context.target_fixture is not None:
>         inject_fixture(request, context.target_fixture, return_value)
> 
>     request.config.hook.pytest_bdd_after_step(**kw)
> ```
> 
> If this helps :)

Great code! it's worked in my case.
u can changed this pr

**Comment by jsa34** [comment]
@youtux are you happy with the codecov?  If so, can we merge?

**Review by jsa34 — APPROVED** [review]

File: src/pytest_bdd/scenario.py [hunk 1]
```diff
@@ -201,6 +202,27 @@ def _execute_step_function(
...
+        return {k: v for k, v in parsed_args.items() if k in func_sig.parameters}
+
+    def _get_argument_values(kwargs: dict) -> dict:
+        """Get default values or request fixture values for missing arguments."""
+        for arg in get_args(context.step_func):
+            if arg not in kwargs:
+                param = func_sig.parameters.get(arg)
+                if param:
+                    if param.default != param.empty:
+                        kwargs[arg] = param.default
```
Comment (youtux) [inline src/pytest_bdd/scenario.py:221]: this is indeed unreachable code

File: src/pytest_bdd/scenario.py [hunk 2]
```diff
@@ -210,24 +214,26 @@ def _execute_step_function(
         "step_func_args": {},
     }
     request.config.hook.pytest_bdd_before_step(**kw)
-    args = get_args(context.step_func)
 
     try:
-        kwargs = parse_step_arguments(step=step, context=context)
+        # Use internal methods without passing redundant arguments
+        parsed_args = parse_step_arguments(step=step, context=context)
+        kwargs = {k: v for k, v in parsed_args.items() if k in func_sig.parameters}
```
Comment (youtux) [inline src/pytest_bdd/scenario.py:221]: is this filtering required? can't we use `kwargs = parsed_args`?

**Comment by youtux** [comment]
Thank you for your contribution @yunxuo. I took the liberty of applying fixes as I saw fit. 

@jsa34 can you review again please?

**Comment by jsa34** [comment]
Looks good to me - thanks for revisiting it for us :)

**Review by jsa34 — APPROVED** [review]

**Pull request merged**